packages/openapi-v3/src/consolidate-schema.ts

```diff
@@ -0,0 +1,106 @@
+import compare from 'json-schema-compare';
+import _ from 'lodash';
+import {
+  ISpecificationExtension,
+  isSchemaObject,
+  OpenApiSpec,
+  PathsObject,
+  ReferenceObject,
+  SchemaObject,
+} from './types';
+
+let $schemaRefs: {
+  [schema: string]: SchemaObject | ReferenceObject;
+};
+let $paths: PathsObject;
+
+/*
+ *  Recursively search OpenApiSpec PathsObject for SchemaObjects with title property.
+ *  Move reusable schema bodies to #/components/schemas and replace with json pointer.
+ *  Handles collisions /w title, schema body pair comparision.
+ */
+export function ConsolidateSchemaObjects(spec: OpenApiSpec): OpenApiSpec {
+  $schemaRefs =
+    spec.components && spec.components.schemas
+      ? _.cloneDeep(spec.components.schemas)
+      : {};
+  $paths = _.cloneDeep(spec.paths);
+
+  recursiveWalk($paths);
+
+  const updatedSpec = {
+    ...spec,
+    ...{
+      paths: $paths,
+      components: {...spec.components, ...{schemas: $schemaRefs}},
+    },
+  };
+
+  // tidy up empty objects
+  if (Object.keys(updatedSpec.components.schemas).length === 0) {
+    delete updatedSpec.components.schemas;
+  }
+  if (Object.keys(updatedSpec.components).length === 0) {
+    delete updatedSpec.components;
+  }
+
+  return updatedSpec;
+}
+
+function recursiveWalk(rootSchema: ISpecificationExtension) {
+  if (rootSchema !== null && typeof rootSchema == 'object') {
+    Object.entries(rootSchema).map(([key, subSchema]) => {
```

Comment:
`map` is typically used to transform values (`['a', 'b'] -> ['c', 'd']`). Since your callback is not returning any value, you should use `forEach` instead.

```suggestion
    Object.entries(rootSchema).forEach(([key, subSchema]) => {
```